src/Discord.Net.SlashCommands/TypeConverters/DefaultEntityTypeConverter.cs

```diff
@@ -0,0 +1,37 @@
+using Discord.WebSocket;
+using System;
+using System.Threading.Tasks;
+
+namespace Discord.SlashCommands
+{
+    internal abstract class DefaultEntityTypeConverter<T> : TypeConverter<T> where T : class
+    {
+        public override Task<TypeConverterResult> ReadAsync (ISlashCommandContext context, SocketSlashCommandDataOption option, IServiceProvider services)
+        {
+            if (option.Value != null)
+                return Task.FromResult(TypeConverterResult.FromSuccess(option.Value as T));
+            else
+                return Task.FromResult(TypeConverterResult.FromError(SlashCommandError.ParseFailed, $"Provided input cannot be read as {nameof(IChannel)}"));
+        }
+    }
+
+    internal class DefaultRoleConverter<T> : DefaultEntityTypeConverter<T> where T : class, IRole
```

Comment:
This can be replaced with a simple switch with types

Reply:
using the TypeConverter pattern makes this overrideable by users